KS-11/Adding restaurant login and register page

shrijan00003 opened the pull request:
- created a simple login page and register page for restaurant login and register

shrijan00003 committed:
methods added on http api
added on request api
api urls made not used yet
modal is created with useModal custom hooks
login page will be using usemodal
auth-layout is created

shrijan00003 committed:
rebase conflict resolved

shrijan00003 committed:
login page created for admin login
storage created
formik added

shrijan00003 committed:
conflict resolved

shrijan00003 committed:
yup added for validation
spinner svg is added
input element created
modal element with hook is created
login view is created
validdation added

shrijan00003 committed:
import fixed

shrijan00003 committed:
yup validation for password added

shrijan00003 committed:
register added with validation and routing

shrijan00003 committed:
optional chaining in response interceptor

shrijan00003 committed:
server error is added to confirmPassword error for now

shrijan00003 committed:
language file added and tested

akasrai commented on src/api/request.api.ts:
```diff
@@ -1,9 +1,26 @@
 };
+
+export const adminLogin = async (payload: { email: string; password: string }) => {
+  const { data, error } = await http.post(`/auth/signin`, payload);
+  if (!error && data?.token) {
+    setUserDetails(data);
+  }
```
We can do this by creating a reducer `useReducer` hook. Say, this hooks handles the `AuthContext` in every update in `AuthContext` it will also update into `storage` 

```js

export const reducer = (
  state: AuthState = initialState,
  action: Action,
): any => {
  switch (action.type) {
    case SIGN_IN_PENDING:
      return updateObject(state, {
        isSigningIn: true,
      });

    case SIGN_IN_SUCCESS:
      storage.set('auth', action.payload);

      return updateObject(state, {
        isSigningIn: false,
        isAuthenticated: true,
        user: action.payload.user,
        token: action.payload.token,
        roles: action.payload.roles,
      });

    case SIGN_IN_ERROR:
      return updateObject(state, {
        isSigningIn: false,
      });

    case SIGN_OUT:
      storage.remove('auth');

      return updateObject(state, {
        ...initialState,
      });

    case RESTORE_AUTH:
      return updateObject(state, {
        isSigningIn: false,
        isAuthenticated: true,
        ...action.payload,
      });

    default:
      return state;
  }
};
```

Here, this reducer handles all the states in `AuthContext`. We use authContext  to store current user details bc we gonna need it very often. LocalStorage will be just to persist the `AuthContext` so we can regain the state of user when browser is refreshed. So now this reducer can be used in auth module as 

```js
const signIn = async (dispatch: Function, loginRequest: Loginrequest) => {
  dispatch({type: auth.SIGN_IN_PENDING});
    const {data, error} = await adminLogin(loginRequest);

    if (error) {
      dispatch({type: auth.SIGN_IN_ERROR});
      return handleSignInError(VALIDATION.ERROR_IN_SIGNIN);
    }

    dispatch({
      type: auth.SIGN_IN_SUCCESS,
      payload: {user: googleSignIn.user, token: data.token, roles: data.roles},
    });
};


const App = () => {
  const [checkAuth, setCheckAuth] = useState<boolean>(true);
  const {setCurrentAuth, isSigningIn} = useContext(AuthContext);
  const [authState, dispatch] = useReducer(reducer, initialState);

  useEffect(() => {
    setCurrentAuth(authState);

   // this is when user refresh the page
    if (checkAuth) {
      setCheckAuth(false);
      restoreAuthentication(dispatch);
    }
  }, [authState]);

  if (isSigningIn) {
    return <Loader />;
  }

  return (
    .....
    <Button
      onCLick={() => signIn(dispatch, loginDetails)}
    />
  );
};

const restoreAuthentication = async (dispatch: Function) => {
  dispatch({type: auth.SIGN_IN_PENDING});
  const {data}: ApiResponse = await storage.get('auth');

  if (data) {
    return dispatch({type: auth.RESTORE_AUTH, payload: data});
  }

  dispatch({type: auth.SIGN_IN_ERROR});
};
```
Well, this comment is getting too long and messed up. We can have a discussion on this.

shrijan00003 replied:
I was about to ask about this, this is the expected comment :)

akasrai commented on src/app/app.storage.ts:
```diff
@@ -0,0 +1,36 @@
+
+const l = localStorage || window.localStorage;
+
+export const getItem = (key: StorageKey) => {
+  return l.getItem(key);
+};
+
+export const setItem = (key: StorageKey, item: string) => {
+  l.setItem(key, item);
+};
+
+export const setUserDetails = (data: any) => {
+  const serializedData = serialize(data);
+
+  setItem(StorageKey.USER_DETAILS, serializedData);
+};
+
+export const getAccessToken = () => {
+  const localData = getItem(StorageKey.USER_DETAILS);
+  const parsedDetails: any = parse(localData || '');
+
+  return parsedDetails?.token || null;
+};
+
+export const changeAccessToken = (token: string) => {
+  const localData = getItem(StorageKey.USER_DETAILS);
+  if (!localData) {
+    const serializedToken = serialize({ token });
+    setItem(StorageKey.USER_DETAILS, serializedToken);
+  } else {
+    const userDetails = { ...parse(localData), token };
+    setItem(StorageKey.USER_DETAILS, serialize(userDetails));
+  }
+};
```
these kinds of helper methods will always be confusing. For example we use them as
```js
import {setItem, getItem } = from 'app.storage,ts';

function any() {
      const token = getItem('token');
      const token2 = setItem('token','dadadaarweweq');
}
```
Now, when code base grows you gonna freak out what the hell is this `setItem`. You need to looka all the way up to imported file to check what its actually doing. So, for these kinda of helper I would prefer to keep in helper module as follows

```js
const hasLocalStorage = localStorage || window.localStorage;
export const storage = {
  get(key) {
    const { error, data } = parse(localStorage.getItem(key));

    if (error) {
      return withError(error);
    }

    if (!hasLocalStorage) {
      return withError(MESSAGE.NO_LOCAL_STORAGE);
    }

    return withData(data);
  },

  set(key, value) {
    if (!hasLocalStorage) {
      return withError(MESSAGE.NO_LOCAL_STORAGE);
    }

    return withData(localStorage.setItem(key, serialize(value)));
  },

  clear(key = null) {
    if (!hasLocalStorage) {
      return withError(MESSAGE.NO_LOCAL_STORAGE);
    }

    localStorage.getItem(key)
      ? localStorage.removeItem(key)
      : localStorage.clear();

    return withData(localStorage.getItem(key));
  }
};
}

```
All the possible exception are handled right away. And use case can be
```js
   import {storage} from 'local-storage.helper.ts';

   storage.set("token",token);
   const {data, error} = storage.getToken('token');
   storage.clear();
``` 
Doesn't it looks more clean and convenient? We can know by seeing the code that token is being stored in storage or being fetched from storage. If you agree we can go this way.

shrijan00003 replied:
loved the one with getter and setter :) 👍

akasrai commented on src/auth/auth.types.ts:
```diff
@@ -0,0 +1,11 @@
+  password: string;
+}
+
+export interface RegisterPayload {
```
I liked the naming prefixed with `Payload` 👍

akasrai commented on src/auth/views/login.tsx:
```diff
@@ -0,0 +1,93 @@
+  const { t } = useTranslation();
+  const loginModal = useModal(true);
+
+  const initialValues: LoginPayload = {
+    email: '',
+    password: '',
+  };
```
These kinda variables can be placed outside the Component because every time Login re-renders a new object is created and this initial value has nothing to do with this because we never gonna update their values .

akasrai commented on src/auth/views/login.tsx:
```diff
@@ -0,0 +1,93 @@
+            const isValid = () => props.isValid && Object.values(props.values).some(Boolean);
+
+            return (
+              <form onSubmit={props.handleSubmit}>
+                <Input
+                  title={at('EMAIL')}
+                  name='email'
+                  type='email'
+                  placeholder={at('ENTER_EMAIL')}
+                  error={!!error('email')}
+                  helperText={helpText('email')}
+                  handleChange={props.handleChange}
+                  handleOnBlur={props.handleBlur}
+                />
+                <Input
+                  name='password'
+                  type='password'
+                  title={at('PASSWORD')}
+                  helperText={helpText('password')}
+                  placeholder={at('ENTER_PASSWORD')}
+                  error={!!error('password')}
+                  handleOnBlur={props.handleBlur}
+                  handleChange={props.handleChange}
+                />
+                <SubmitButton loading={props.isSubmitting} text={at('LOGIN')} disabled={!isValid()} />
+              </form>
```
This return method is getting too long. These JSX can be separated into a new Component within the file.

akasrai commented on src/auth/views/register.tsx:
```diff
@@ -0,0 +1,108 @@
+  const registerModal = useModal(true);
+  const { t } = useTranslation();
+
+  const initialValues: RegisterPayload = {
+    name: '',
+    email: '',
+    password: '',
+    confirmPassword: '',
+  };
```
Let's move Outside the Component. As mentioned above these helps just to increase the number of lines of method. We can always try to make method as small as possible.

akasrai commented on src/auth/views/register.tsx:
```diff
@@ -0,0 +1,108 @@
+  const at = (text: string) => t(`auth.${text}`);
+
+  return (
+    <AuthLayout>
```
JSX can be simplified into smaller components

akasrai reviewed (changes requested):
For me we should start the refactoring and better implementation from the beginning of project because we neither gonna refactor them in the future nor we can develop a new module. So, I hv commented some of my thoughts we can further discuss on these and  refactor the codes.

shrijan00003 committed:
app storage modified to have getter and setter
, object with key is added
, errors in language file is added

shrijan00003 committed:
react test renderer added,
app.test.tsx removed for now
other test are fixed

shrijan00003 committed:
methods added to storage objectare tested

shrijan00003 committed:
storage test added
auth context is added
login and register page is refactored to have form seperately

shrijan00003 committed:
auth service file is added

shrijan00003 commented:
@akasrai  please review now and let me know if anything else have to be fixed.

akasrai reviewed (approved):
Looks better 👍👍
